### PR TITLE
create a scratchpad untitled workingcopy for the IW model

### PR DESCRIPTION
**Pull request by amunger**
for https://github.com/microsoft/vscode/issues/174485

### CODE REVIEW

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts [hunk 1]
```diff
@@ -121,6 +121,7 @@ export class UntitledFileWorkingCopy<M extends IUntitledFileWorkingCopyModel> ex
 		readonly resource: URI,
 		readonly name: string,
 		readonly hasAssociatedFilePath: boolean,
+		readonly scratchPad: boolean,
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts:124]:
```suggestion
		private readonly scratchPad: boolean,
```

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts [hunk 2]
```diff
@@ -130,6 +131,8 @@ export class UntitledFileWorkingCopy<M extends IUntitledFileWorkingCopyModel> ex
 	) {
 		super();
 
+		this.capabilities = scratchPad ? WorkingCopyCapabilities.Untitled | WorkingCopyCapabilities.Scratchpad : WorkingCopyCapabilities.Untitled;
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts:134]: You can still do this from the property in line 95.

**Review by bpasero — COMMENTED** [review]
I think I like the direction of the new capability and how it can be applied via the manager. But applying the capability needs to be done with care in all locations: sometimes we use the `onDidChangeDirty` event to update an indicator but then sometimes we use the `dirtyCount` or `dirtyWorkingCopies` or `hasDirty`. Since we are not disabling the dirty state for scratchpad working copies (which I think is good so that backups are triggered) we probably have to jump into each location and figure out how to treat scratchpad working copies...

**Comment by amunger** [comment]
I went through each usage of the `dirtyWorkingCopies` and `dirtyCount` getters and I'm pretty confident that they currently work as is with a couple exceptions:

`saveAll` after `bulkEdit` in `bulkEditService`
  untitled files aren't affected? (bug https://github.com/microsoft/vscode/issues/181576)

`doSaveAllBeforeShutdown` in the `WorkingCopyBackupTracker` would include scratchpad editors, except that method is never called with `untitled: true`
I'm also fuzzy about editor capabilities vs workingCopy capabilities and if I need to make changes there as well.

**Comment by amunger** [comment]
I'm also up for adding some tests - it looks like the `WorkingCopyBackupTracker.tests` doesn't currently run tests on untitled working copies, but I could try adding some in unless there is a more appropriate place to put those.

**Comment by bpasero** [comment]
@amunger I did a pass over every use of dirty state tracking for working copies and I am now even more convinced that the best way forward is to exclude scratchpad working copies from dirty tracking entirely, meaning:
* no dirty change event is emitted and they never report as dirty
* they are excluded from the various working copy service methods that allow to get dirty count or dirty working copies (this likely comes for free given the first point)

As soon as we do that, scrapbook working copies will not cause dirty indicators to appear, nor confirmation dialogs to appear, nor will any of the save actions trigger a save.

However, there are a few locations where we explicitly have to account for scrapbooks that have content. Here are the ones I found:

**Editor Group View**
We automatically turn a preview editor into non-preview if it is dirty ([here](https://github.com/microsoft/vscode/blob/424388032a508f0c330e2aa07198941687597c76/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L987)). Otherwise, preview editors will automatically close and we probably do not want that for scratchpad that have content.

**Editor Observer**
Similar to the above, our editor observer will automatically close non-dirty editors based on user settings, if a maximum of editors is configured ([here](https://github.com/microsoft/vscode/blob/424388032a508f0c330e2aa07198941687597c76/src/vs/workbench/browser/parts/editor/editorsObserver.ts#L341)). We cannot allow that for scratchpads with content I think, because you would loose your data easily.

**Working Copy Backup Tracker**
This is the big one that I had mentioned already: dirty state is used to track backups in various locations. We will probably have to change all the usages to a new concept that:
* includes everything we have today for dirty working copies
* includes scratchpads that have content

For that we need something new that is not "dirty" but signals the intent. I am open for suggestions.

File: src/vs/workbench/contrib/files/common/dirtyFilesIndicator.ts [hunk 3]
```diff
@@ -42,6 +42,7 @@ export class DirtyFilesIndicator extends Disposable implements IWorkbenchContrib
 
 	private onWorkingCopyDidChangeDirty(workingCopy: IWorkingCopy): void {
 		const gotDirty = workingCopy.isDirty();
+
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/common/dirtyFilesIndicator.ts:45]: Unrelated change

File: src/vs/workbench/contrib/files/common/dirtyFilesIndicator.ts [hunk 4]
```diff
@@ -53,7 +54,6 @@ export class DirtyFilesIndicator extends Disposable implements IWorkbenchContrib
 
 	private updateActivityBadge(): void {
 		const dirtyCount = this.lastKnownDirtyCount = this.workingCopyService.dirtyCount;
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/common/dirtyFilesIndicator.ts:56]: Unrelated change

File: src/vs/workbench/common/editor/editorInput.ts [hunk 5]
```diff
@@ -180,6 +180,14 @@ export abstract class EditorInput extends AbstractEditorInput {
 		return false;
 	}
 
+	/**
+	 * Returns if the input has changes that are not backed up,
+	 * whether or not it is marked dirty.
+	 */
+	hasUnsavedContent(): boolean {
```
Comment (bpasero) [inline src/vs/workbench/common/editor/editorInput.ts:187]: I have mixed feelings about this one. We could simply introduce a new `EditorInputCapability` called the same as the working copy (`Scratchpad`) and then check on the capability instead?

File: src/vs/workbench/services/workingCopy/common/workingCopy.ts [hunk 6]
```diff
@@ -21,7 +21,13 @@ export const enum WorkingCopyCapabilities {
 	 * additional input when saving, e.g. an
 	 * associated path to save to.
 	 */
-	Untitled = 1 << 1
+	Untitled = 1 << 1,
+
+	/**
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopy.ts:26]: This needs a better explainer, esp. "dirty" is no longer true for scratchpads.

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts [hunk 7]
```diff
@@ -121,6 +121,7 @@ export class UntitledFileWorkingCopy<M extends IUntitledFileWorkingCopyModel> ex
 		readonly resource: URI,
 		readonly name: string,
 		readonly hasAssociatedFilePath: boolean,
+		private readonly scratchPad: boolean,
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts:124]:
```suggestion
		private readonly isScratchpad: boolean,
```

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts [hunk 8]
```diff
@@ -190,7 +192,9 @@ export class UntitledFileWorkingCopy<M extends IUntitledFileWorkingCopyModel> ex
 		await this.doCreateModel(untitledContents);
 
 		// Untitled associated to file path are dirty right away as well as untitled with content
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts:194]: Should probably update the comment here

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts [hunk 9]
```diff
@@ -91,6 +91,8 @@ export interface INewOrExistingUntitledFileWorkingCopyOptions extends INewUntitl
 	 * Note: the resource will not be used unless the scheme is `untitled`.
 	 */
 	untitledResource: URI;
+
+	scratchPad?: boolean;
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts:95]:
```suggestion
	isScratchpad?: boolean;
```

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts [hunk 10]
```diff
@@ -91,6 +91,8 @@ export interface INewOrExistingUntitledFileWorkingCopyOptions extends INewUntitl
 	 * Note: the resource will not be used unless the scheme is `untitled`.
 	 */
 	untitledResource: URI;
+
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts:94]: Comments?

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts [hunk 11]
```diff
@@ -167,6 +169,7 @@ export class UntitledFileWorkingCopyManager<M extends IUntitledFileWorkingCopyMo
 		// Handle untitled resource
 		else if (options.untitledResource?.scheme === Schemas.untitled) {
 			massagedOptions.untitledResource = options.untitledResource;
+			massagedOptions.scratchPad = options.scratchPad;
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts:172]: I wonder: why not allow scratchpads even with associated resource? You could then save a scratchpad without asking the user for a path?

**Review by bpasero — CHANGES_REQUESTED** [review]
Did a quick review

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts (hunk 11 above)
Comment (amunger) [reply src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts:172]: they would just be titled `untitled-n` in that case, so we would probably need a way to differentiate them in the UI from non-scratchpad untitled files. It feels simplest to just leave that out for now.

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts (hunk 8 above)
Comment (amunger) [reply src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopy.ts]: The check now only exists within `setDirty` to make the usage consistent in the file, but I added a comment in there.

File: src/vs/workbench/common/editor/editorInput.ts (hunk 5 above)
Comment (amunger) [reply src/vs/workbench/common/editor/editorInput.ts:187]: that sounds fine - I was thinking that we wanted information about if there was any content as well, but it looks like we can just get away with checking the capability on its own for now

File: src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts (hunk 11 above)
Comment (bpasero) [reply src/vs/workbench/services/workingCopy/common/untitledFileWorkingCopyManager.ts]: 👍

File: src/vs/workbench/common/editor/editorInput.ts (hunk 5 above)
Comment (bpasero) [reply src/vs/workbench/common/editor/editorInput.ts]: That is a good point but I think its also fine like now.

**Review by bpasero — CHANGES_REQUESTED** [review]
I like all the changes but think we now need to focus on backup tracker. I am not convinced we have all the changes yet to ensure backups are present, including shutdown.

Backup tracker makes extensive use of tracking "dirty" working copies. Scratchpad working copies are never dirty. You schedule a backup in `onDidChangeContent` which is important, but there are other places that are not tracked. Can you make a pass and see what to do? I think we will have to introduce a new field on the working copy service and maybe on the working copy itself to signal back that content has changed in the past and that a backup should be done.

I think you only need to check all the (3?) `workingCopyBackupTracker.ts` files for this.

**Comment by amunger** [comment]
oof, yes I missed that since the IW does it's own backup at the moment (hopefully can integrate as a next step)

common
schedule backup when registered with Backup Tracker and dirty - *update* should also include scratchpads with content
schedule backup on Dirty Change - won't trigger for scratchpads, should be covered by `onContentChange` except for dropping the backups
schedule backup on Content Change - applies to scratchpad, *update* discard the backup when un-modified since we miss that from `onDirtyChange`

web
onFinalBeforeShutdown - *update* scratchpads that failed backup should veto
  without hot-exit, scratchpads will be discarded by design

electron-sandbox
handleDirtyBeforeShutdown - *update* should include scratchpads with content
  - if backups fail, we should veto shutdown
  - without backups, scratchpads will be discarded by design

File: src/vs/workbench/services/textmodelResolver/test/browser/textModelResolverService.test.ts [hunk 12]
```diff
@@ -157,6 +157,7 @@ suite('Workbench - TextModelResolverService', () => {
 	test('resolve untitled', async () => {
 		const service = accessor.untitledTextEditorService;
 		const untitledModel = service.create();
+		untitledModel.isModified = untitledModel.isDirty;
```
Comment (bpasero) [inline src/vs/workbench/services/textmodelResolver/test/browser/textModelResolverService.test.ts:160]: This change seems unintentional?

File: src/vs/workbench/services/textfile/common/textFileEditorModel.ts [hunk 13]
```diff
@@ -674,6 +674,10 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
 		return this.dirty;
 	}
 
+	isModified(): boolean {
+		return this.dirty;
```
Comment (bpasero) [inline src/vs/workbench/services/textfile/common/textFileEditorModel.ts:678]: I would prefer to call `this.isDirty()` so that we make sure the 2 are always in sync.

File: src/vs/workbench/services/textfile/common/textFileEditorModel.ts [hunk 14]
```diff
@@ -674,6 +674,10 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
 		return this.dirty;
 	}
 
+	isModified(): boolean {
```
Comment (bpasero) [inline src/vs/workbench/services/textfile/common/textFileEditorModel.ts:677]: We should add this to the tests for text file model to assert it works.

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 15]
```diff
@@ -67,7 +67,7 @@ export interface IUntitledTextEditorModel extends ITextEditorModel, ILanguageSup
 	resolve(): Promise<void>;
 }
 
-export class UntitledTextEditorModel extends BaseTextEditorModel implements IUntitledTextEditorModel {
+export class UntitledTextEditorModel extends BaseTextEditorModel implements IUntitledTextEditorModel, IWorkingCopy {
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:70]: Not needed, `IUntitledTextEditorModel` already implements `IWorkingCopy`?

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 16]
```diff
@@ -243,6 +243,10 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
 		return this.dirty;
 	}
 
+	isModified(): boolean {
+		return this.dirty;
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:247]: I would prefer `this.isDirty()`

File: src/vs/workbench/test/common/workbenchTestServices.ts [hunk 17]
```diff
@@ -176,7 +176,7 @@ export class TestWorkingCopy extends Disposable implements IWorkingCopy {
 	private readonly _onDidSave = this._register(new Emitter<IStoredFileWorkingCopySaveEvent>());
 	readonly onDidSave = this._onDidSave.event;
 
-	readonly capabilities = WorkingCopyCapabilities.None;
+	readonly capabilities: WorkingCopyCapabilities = WorkingCopyCapabilities.None;
```
Comment (bpasero) [inline src/vs/workbench/test/common/workbenchTestServices.ts:179]: Not needed, TS will infer the type from the value.

File: src/vs/workbench/test/common/workbenchTestServices.ts [hunk 18]
```diff
@@ -203,6 +203,10 @@ export class TestWorkingCopy extends Disposable implements IWorkingCopy {
 		return this.dirty;
 	}
 
+	isModified(): boolean {
+		return this.dirty;
```
Comment (bpasero) [inline src/vs/workbench/test/common/workbenchTestServices.ts:207]: I would prefer `this.isDirty()`

File: src/vs/workbench/services/workingCopy/test/browser/untitledFileWorkingCopy.test.ts [hunk 19]
```diff
@@ -102,6 +102,7 @@ suite('UntitledFileWorkingCopy', () => {
 			uri,
 			basename(uri),
 			hasAssociatedFilePath,
+			false,
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/test/browser/untitledFileWorkingCopy.test.ts:105]: We should also have a test here for scratchpads to ensure the manager passes the flag on accordingly.

File: src/vs/workbench/services/workingCopy/common/workingCopyService.ts [hunk 20]
```diff
@@ -265,6 +271,10 @@ export class WorkingCopyService extends Disposable implements IWorkingCopyServic
 		return this.workingCopies.filter(workingCopy => workingCopy.isDirty());
 	}
 
+	get modifiedWorkingCopies(): IWorkingCopy[] {
+		return this.workingCopies.filter(workingCopy => workingCopy.isModified);
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyService.ts:275]: `isModified` is a function, not a property

File: src/vs/workbench/services/workingCopy/common/workingCopyService.ts [hunk 21]
```diff
@@ -68,6 +68,12 @@ export interface IWorkingCopyService {
 	 */
 	readonly dirtyWorkingCopies: readonly IWorkingCopy[];
 
+	/**
+	 * All working copies with unsaved changes,
+	 * including scratchpads, which are never dirty
+	 */
+	readonly modifiedWorkingCopies: readonly IWorkingCopy[];
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyService.ts:75]: I like this, also needs a unit test

File: src/vs/workbench/services/workingCopy/common/resourceWorkingCopy.ts [hunk 22]
```diff
@@ -161,4 +170,5 @@ export abstract class ResourceWorkingCopy extends Disposable implements IResourc
 	abstract revert(options?: IRevertOptions): Promise<void>;
 
 	//#endregion
+
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/resourceWorkingCopy.ts:173]: This formatting change is not needed.

File: src/vs/workbench/services/workingCopy/common/resourceWorkingCopy.ts [hunk 23]
```diff
@@ -143,6 +143,15 @@ export abstract class ResourceWorkingCopy extends Disposable implements IResourc
 
 	//#endregion
 
+	//#region Modified Tracking
+
+	// Overridden when the Working Copies need to differentiate between
+	// dirty and modified e.g. scratchpads
+	isModified(): boolean {
+		return this.isDirty();
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/resourceWorkingCopy.ts:151]: Do we still need this when we adopt `isModified()` to return `this.isDirty()` in most places?

File: src/vs/workbench/services/workingCopy/browser/workingCopyBackupTracker.ts [hunk 24]
```diff
@@ -32,27 +32,27 @@ export class BrowserWorkingCopyBackupTracker extends WorkingCopyBackupTracker im
...
+		// As such we need to check sync if there are any modified working
 		// copies that have not been backed up yet and then prevent the
 		// shutdown if that is the case.
 
-		const dirtyWorkingCopies = this.workingCopyService.dirtyWorkingCopies;
-		if (!dirtyWorkingCopies.length) {
-			return false; // no dirty: no veto
+		const modifiedWorkingCopies = this.workingCopyService.modifiedWorkingCopies;
+		if (!modifiedWorkingCopies.length) {
+			return false; // no unsaved changes: no veto
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/browser/workingCopyBackupTracker.ts:41]: The comments refer to "unsaved changes" but I think for better understanding can we refer to this as "modified"?

File: src/vs/workbench/services/workingCopy/browser/workingCopyBackupTracker.ts [hunk 25]
```diff
@@ -32,27 +32,27 @@ export class BrowserWorkingCopyBackupTracker extends WorkingCopyBackupTracker im
...
 		}
 
-		if (!this.filesConfigurationService.isHotExitEnabled) {
+		if (this.workingCopyService.hasDirty && !this.filesConfigurationService.isHotExitEnabled) {
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/browser/workingCopyBackupTracker.ts:44]: Here you are using `hasDirty`, but I think we need `hasModified` and still veto when hot exit is disabled because you will loose the scratchpad otherwise.

File: src/vs/workbench/services/workingCopy/common/workingCopyBackupTracker.ts [hunk 26]
```diff
@@ -159,13 +159,16 @@ export abstract class WorkingCopyBackupTracker extends Disposable {
 			return;
 		}
 
-		// Schedule backup if dirty
-		if (workingCopy.isDirty()) {
+		// Schedule backup for unsaved content
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyBackupTracker.ts:162]: "for modified working copies"

File: src/vs/workbench/services/workingCopy/common/workingCopyBackupTracker.ts [hunk 27]
```diff
@@ -159,13 +159,16 @@ export abstract class WorkingCopyBackupTracker extends Disposable {
...
+		if (workingCopy.isModified()) {
 			// this listener will make sure that the backup is
 			// pushed out for as long as the user is still changing
 			// the content of the working copy.
 			this.scheduleBackup(workingCopy);
 		}
+		else {
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyBackupTracker.ts:169]: This sounds like a dangerous change to me, why?

File: src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts [hunk 28]
```diff
@@ -68,21 +68,21 @@ export class NativeWorkingCopyBackupTracker extends WorkingCopyBackupTracker imp
...
 			// Dirty working copies need treatment on shutdown
-			const dirtyWorkingCopies = this.workingCopyService.dirtyWorkingCopies;
-			if (dirtyWorkingCopies.length) {
-				return await this.onBeforeShutdownWithDirty(reason, dirtyWorkingCopies);
+			const modifiedWorkingCopies = this.workingCopyService.modifiedWorkingCopies;
+			if (modifiedWorkingCopies.length) {
+				return await this.onBeforeShutdownWithModified(reason, modifiedWorkingCopies);
 			}
 
 			// No dirty working copies
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts:76]: "No modified working copies"

File: src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts [hunk 29]
```diff
@@ -97,40 +97,38 @@ export class NativeWorkingCopyBackupTracker extends WorkingCopyBackupTracker imp
 			}
 
 			// If we still have dirty working copies, we either have untitled ones or working copies that cannot be saved
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts:99]: "have modified working copies"

File: src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts [hunk 30]
```diff
@@ -97,40 +97,38 @@ export class NativeWorkingCopyBackupTracker extends WorkingCopyBackupTracker imp
...
-			const remainingDirtyWorkingCopies = this.workingCopyService.dirtyWorkingCopies;
-			if (remainingDirtyWorkingCopies.length) {
-				return this.handleDirtyBeforeShutdown(remainingDirtyWorkingCopies, reason);
+			const remainingModifiedWorkingCopies = this.workingCopyService.modifiedWorkingCopies;
+			if (remainingModifiedWorkingCopies.length) {
+				return this.handleModifiedBeforeShutdown(remainingModifiedWorkingCopies, reason);
 			}
 
-			return this.noVeto([...dirtyWorkingCopies]); // no veto (dirty auto-saved)
+			return this.noVeto([...modifiedWorkingCopies]); // no veto (dirty auto-saved)
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts:105]: "modified auto-saved"

File: src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts [hunk 31]
```diff
@@ -208,27 +208,25 @@ export class NativeWorkingCopyBackupTracker extends WorkingCopyBackupTracker imp
 	}
 
 	private showErrorDialog(msg: string, workingCopies: readonly IWorkingCopy[], error?: Error): void {
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts:210]: Why was the line removed?

File: src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts [hunk 32]
```diff
@@ -139,14 +137,16 @@ export class NativeWorkingCopyBackupTracker extends WorkingCopyBackupTracker imp
...
+			const remainingModifiedWorkingCopies = modifiedWorkingCopies.filter(workingCopy => !backups.includes(workingCopy));
+			this.showErrorDialog(localize('backupTrackerBackupFailed', "The following editors with unsaved changes could not be saved to the back up location."), remainingModifiedWorkingCopies, backupError);
 
 			return true; // veto (the backup failed)
 		}
 
 		// Since a backup did not happen, we have to confirm for
 		// the working copies that did not successfully backup
 
+		const remainingDirtyWorkingCopies = modifiedWorkingCopies.filter(workingCopy => !backups.includes(workingCopy) && workingCopy.isDirty());
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/electron-sandbox/workingCopyBackupTracker.ts:149]: Shouldnt we also warn about scratchpads not getting backed up?

**Review by bpasero — CHANGES_REQUESTED** [review]
I really like the term "modified" to signal backups are needed for a working copy. But then we should also make sure that when we have a comment, we refer to "modified" as term and not "dirty" or "unsaved changes". Makes it easier to understand the code imho 👍 

I left lots of little review comments but on a high level think that you missed some cases where a modified working copy without backup should still block the shutdown from happening. If a user has a modified scrapbook and hot exit is disabled we still need to show a dialog, to not loose user data.

Overall I think we are getting close to done.

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts (hunk 15 above)
Comment (amunger) [reply src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:70]: It made the error more understandable when I was missing properties, but I can take it back out if that's preferred

File: src/vs/workbench/test/common/workbenchTestServices.ts (hunk 17 above)
Comment (amunger) [reply src/vs/workbench/test/common/workbenchTestServices.ts:179]: another phantom error, yesterday it was complaining that capabilities was of _type_ `WorkingCopyCapabilities.None` so I couldn't override it. Seems fine now tho

File: src/vs/workbench/services/workingCopy/common/workingCopyBackupTracker.ts (hunk 27 above)
Comment (amunger) [reply src/vs/workbench/services/workingCopy/common/workingCopyBackupTracker.ts:169]: yes I could see some unintended consequences from this. The reason was that we only discard backups in the onDidChangeDirty, which scratchpads will never trigger, meaning that we would never discard scratchpad backups. Is that OK?

**Comment by amunger** [comment]
With the change to block shutdown for modified scratchpads, I had to remove the IWs implementation since we need to think more about what 'save' will do for an IW if we're going to force users to take that action.

File: src/vs/workbench/services/workingCopy/test/browser/untitledScratchpadWorkingCopy.test.ts [hunk 33]
```diff
@@ -0,0 +1,273 @@
...
+import { TestServiceAccessor, workbenchInstantiationService } from 'vs/workbench/test/browser/workbenchTestServices';
+
+export class TestUntitledFileWorkingCopyModelFactory implements IUntitledFileWorkingCopyModelFactory<TestUntitledFileWorkingCopyModel> {
+
+	async createModel(resource: URI, contents: VSBufferReadableStream, token: CancellationToken): Promise<TestUntitledFileWorkingCopyModel> {
+		return new TestUntitledFileWorkingCopyModel(resource, (await streamToBuffer(contents)).toString());
+	}
+}
+
+suite('UntitledScratchpadWorkingCopy', () => {
```
Comment (amunger) [inline src/vs/workbench/services/workingCopy/test/browser/untitledScratchpadWorkingCopy.test.ts:26]: copied from UntitledFileWorkingCopy suite with changes to account for scratchpad behavior.

Comment (bpasero) [reply src/vs/workbench/services/workingCopy/test/browser/untitledScratchpadWorkingCopy.test.ts:26]: You could try to reuse the same code in both suites with a parameter for using scratchpad or not.

File: src/vs/workbench/services/workingCopy/common/resourceWorkingCopy.ts (hunk 23 above)
Comment (bpasero) [reply src/vs/workbench/services/workingCopy/common/resourceWorkingCopy.ts:151]: I think the override is fine but I was wondering about the explicit comment you made there, why do you have it? I think we can remove it, if a subclass wants to change this behaviour, they can.

**Review by bpasero — CHANGES_REQUESTED** [review]
Getting there 👍 

Can you search for `dirty` in `src/vs/workbench/services/workingCopy/common/workingCopyBackupTracker.ts` and review where we still need to deal with modified and not dirty, I think its not done yet.

**Comment by bpasero** [comment]
> With the change to block shutdown for modified scratchpads, I had to remove the IWs implementation since we need to think more about what 'save' will do for an IW if we're going to force users to take that action.

I would expect that I can save a scratchpad when I do this explicitly and it would open a dialog asking where to save to, similar to untitled files.

File: src/vs/workbench/services/workingCopy/test/browser/untitledScratchpadWorkingCopy.test.ts (hunk 33 above)
Comment (amunger) [reply src/vs/workbench/services/workingCopy/test/browser/untitledScratchpadWorkingCopy.test.ts:26]: each test would have its own extracted function right? that feels pretty strange. I was hoping to just do a data driven test, but that's not yet supported with the test runner (but might be soon https://github.com/microsoft/vscode/issues/182788).

Comment (bpasero) [reply src/vs/workbench/services/workingCopy/test/browser/untitledScratchpadWorkingCopy.test.ts:26]: Yeah not a big deal. We do this kind of shared tests when we want to execute the exact same tests in web and desktop, but if the tests are differing, maybe not worth the efford.

**Review by bpasero — APPROVED** [review]
I think this is good to go. I spend some time thinking about 2 somewhat loose ends that are slightly off:
- do we need a `onDidChangeModified` event given we have introduced `modified` as a concept? or should we maybe rename the existing `onDidChangeContent`? Should we drop content-change notion and only use modified?
- if you search for `markDirty` in our sources you find a way for clients to mark a untitled working copy as dirty. should this be renamed to `markModified` or is the correct term as it is now?

I am not sure these 2 bullets warrant for blocking this PR any l longer though, thus approving.
